File: bluetooth test 1/BluetoothManager.m
#import "test-Bridging-Header.h"
#import "test-Swift.h"  // Replace with your actual Swift header name

@interface BluetoothBridge () <BluetoothScanDelegate>
@property (nonatomic, strong) BluetoothManager *manager;
@property (nonatomic, strong) NSDictionary *lastDevice;
@end

@implementation BluetoothBridge

- (instancetype)init {
    self = [super init];
    if (self) {
        _manager = [[BluetoothManager alloc] init];
        _manager.delegate = self;
        NSLog(@"BluetoothBridge initialized with manager: %@", _manager);
    }
    return self;
}

- (void)startBluetoothScan {
    NSLog(@"BluetoothBridge: Starting scan...");
    [_manager startScan];
}

- (void)stopBluetoothScan {
    NSLog(@"BluetoothBridge: Stopping scan...");
    [_manager stopScan];
}

// This method will be called by Swift BluetoothManager
- (void)didDiscoverDeviceWithName:(NSString *)name
                             uuid:(NSString *)uuid
                             rssi:(NSInteger)rssi
                            major:(NSInteger)major
                            minor:(NSInteger)minor {
    
    NSLog(@"BluetoothBridge: Device discovered - Name: %@, UUID: %@, RSSI: %ld", 
          name, uuid, (long)rssi);
    
    self.lastDevice = @{
        @"name": name ?: @"Unknown",
        @"uuid": uuid ?: @"",
        @"rssi": @(rssi),
        @"major": @(major),
        @"minor": @(minor)
    };
    
    NSLog(@"BluetoothBridge: lastDevice set to: %@", self.lastDevice);
}

- (void)connectToDeviceWithUUID:(NSString *)uuid {
    [_manager connectToPeripheralWith:uuid];
}

- (NSDictionary *)getLastDiscoveredDevice {
    NSLog(@"BluetoothBridge: getLastDiscoveredDevice called, returning: %@", self.lastDevice);
    return self.lastDevice;
}

@end

File: bluetooth test 1/BluetoothManager.swift
import Foundation
import CoreBluetooth

@objc protocol BluetoothScanDelegate {
    func didDiscoverDevice(name: String?, uuid: String, rssi: Int, major: Int, minor: Int)
}

@objcMembers
class BluetoothManager: NSObject, CBCentralManagerDelegate, CBPeripheralDelegate {
    var centralManager: CBCentralManager?
    
    weak var delegate: BluetoothScanDelegate?
    var discoveredPeripherals: [UUID: CBPeripheral] = [:]

    override init() {
        super.init()
        centralManager = CBCentralManager(delegate: self, queue: nil)
    }

    func startScan() {
        guard let manager = centralManager else { return }
        if manager.state == .poweredOn {
            manager.scanForPeripherals(withServices: nil, options: nil)
            print("Swift: Scanning started")
        }
    }

    func stopScan() {
        centralManager?.stopScan()
        print("Swift: Scanning stopped")
    }

    func centralManager(_ central: CBCentralManager, didDiscover peripheral: CBPeripheral,
                        advertisementData: [String : Any], rssi RSSI: NSNumber) {
        let uuid = peripheral.identifier.uuidString
        let name = peripheral.name ?? "Unknown"
        discoveredPeripherals[peripheral.identifier] = peripheral

        var major = 0
        var minor = 0

        if let manufacturerData = advertisementData[CBAdvertisementDataManufacturerDataKey] as? Data {
            if manufacturerData.count >= 25 {
                // iBeacon format: bytes 20-21: major, 22-23: minor
                major = Int(manufacturerData[20]) << 8 | Int(manufacturerData[21])
                minor = Int(manufacturerData[22]) << 8 | Int(manufacturerData[23])
            }
        }

        delegate?.didDiscoverDevice(name: name, uuid: uuid, rssi: RSSI.intValue, major: major, minor: minor)
    }

    func centralManagerDidUpdateState(_ central: CBCentralManager) {
        print("Bluetooth state changed: \(central.state.rawValue)")
    }
    @objc(connectToPeripheralWith:)
    func connectToPeripheral(uuidString: String) {
        guard let uuid = UUID(uuidString: uuidString),
              let peripheral = discoveredPeripherals[uuid] else {
            print("Peripheral with UUID \(uuidString) not found")
            return
        }

        centralManager?.connect(peripheral, options: nil)
    }
    func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        print("✅ Connected to peripheral: \(peripheral.name ?? "Unknown")")
        peripheral.delegate = self
        peripheral.discoverServices(nil)  // optional: for characteristic discovery
    }

}
